DOCSP-18830: upgrade from legacy API section

ccho-mongodb opened the pull request:
## Pull Request Info

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-18830
https://jira.mongodb.org/browse/DOCSP-20648

### Snooty build log:
https://workerpool-boxgs.mongodbstitch.com/pages/job.html?collName=queue&jobId=62349ece0ab54725f12133e6

### Docs staging link (requires sign-in on MongoDB Corp SSO):
[Driver landing page summary](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-18830-upgrade-legacy/#upgrade-driver-versions)
[Upgrade page](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-18830-upgrade-legacy/upgrade/)
[Legacy page](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-18830-upgrade-legacy/legacy/)
[FAQ page](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-18830-upgrade-legacy/faq/#how-do-i-connect-to-my-mongodb-instance-with-the-legacy-api-)
[Extended JSON page](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-18830-upgrade-legacy/fundamentals/data-formats/document-data-format-extended-json/#extended-json-formats)

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging and workerpool job links in the PR description updated?

### If your page documents a concept, does it meet the following criteria?

- [ ] Target the [Jasmin persona](https://drive.google.com/file/d/14FbBOLCVxwSP6M9BK4Nz1Ir9tzxT8_02/view)
- [ ] Target the [Lucas persona](https://drive.google.com/file/d/1J2vqJxo7ldv7OP_obA9Q-avf0o_ju4Lk/view)

biniona-mongodb commented on source/index.txt:
```diff
@@ -101,6 +102,13 @@ What's New
+-----------------------
+
+Learn what changes you may need to make to your application to upgrade
+versions or from the Legacy API to the current one in the
```
Issue: I believe "Legacy" in "legacy API" is inconsistently capitalized. I'm not sure which is right, but elsewhere in the PR it is lower case.

```suggestion
versions or from the legacy API to the current one in the
```

ccho-mongodb replied:
Good catch, thought I had updated all of the instances. Initially, I started with capital "L", but noticed the FAQ entries used lowercase, and switched to the latter.

biniona-mongodb reviewed (approved):
LGTM with some small issues and suggestions around things that tripped me up

ccho-mongodb commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+
+When you upgrade your driver version or use multiple versions of the
+driver, if there are any breaking changes between the two versions that you
+have not addressed in the :ref:`Breaking Changes <java-breaking-changes>`
```
Sometime in the past couple months, I discovered that this is not a reliable substitution mechanism (sometimes the parser cannot locate the title and therefore omits it), so I've avoided using it since then.

ccho-mongodb commented on source/index.txt:
```diff
@@ -101,6 +102,13 @@ What's New
+Upgrade Driver Versions
+-----------------------
+
+Learn what changes you may need to make to your application to upgrade
```
This doesn't have to do with authoritative guidance.

There's a chance they don't need to make any changes, so action is only needed for the ones that apply. Assuming the reader definitely needs to make changes may only serve to confuse the reader.

ccho-mongodb commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+  to use new format that minimizes serialization compatibility issues
+  across different versions of the driver.
+
+  Any objects that include the ``ObjectId``, serialized with an older version
+  of the driver are not compatible and the driver throws an exception when
```
I think I inadvertently deleted part of the beginning of the sentence before committing. It should read something like:

"Any objects that include an ObjectId field that were serialized with an older version of the driver are not deserializable by this driver version and later. When the driver encounters an incompatible format, it throws an exception."

Does that make sense?

biniona-mongodb replied:
Yup, the proposed phrasing SGTM

ccho-mongodb commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+The breaking changes in this section are categorized by the driver version that
+introduced them.
+
+If you are upgrading or have multiple applications that share data that use
+different driver versions, address any of the breaking changes between the two
+versions, including ones in the later version. For example, if you are
```
I don't think we should remove the information, but maybe I can reword it to make more sense.

Context: If you consult the details of the JAVA project ticket, the confusion around the breaking change was based on serialization of the data and accessing it using applications that each used different versions of the driver.

Proposed solution: Improve parallel structure in the sentence.

"If you are either upgrading the driver version in your application or plan to run multiple applications that use different driver versions, address any of the breaking changes between the two versions. For example, ..."

Does that make sense?

biniona-mongodb commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+the legacy API to the current API.
+
+When you upgrade your driver version or use multiple versions of the
+driver, if there are any breaking changes between the them that you
```
Issue: May be an extra word here

```suggestion
driver, if there are any breaking changes between them that you
```

biniona-mongodb reviewed (approved):
Nice edits! One small thing that may need fixed, but overall looks great.

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+the legacy API to the current API.
+
+When you upgrade your driver version or use multiple versions of the
+driver, if there are any breaking changes between them that you have not
```
It's not possible to use multiple versions of the driver in the same running application (unless you do some crazy class loader things)

ccho-mongodb replied:
That's not the information I was trying to convey here, so I'm guessing I need it to make it more clear or maybe just omit that info and let the smaller audience to which it could apply infer it.

jyemin replied:
Can you describe more what you were trying to convey?

jyemin replied:
I guess this was removed?

ccho-mongodb replied:
Yes, this was a cross-thread communication:
https://github.com/mongodb/docs-java/pull/212#discussion_r824845911

You mentioned that it could be misconstrued, which I agreed with and removed it. The text that replaced it is here:

https://github.com/mongodb/docs-java/pull/212/files#diff-9c988395fdf1141e52aa9fa946e169ad1d9fb7de62fd44fc1b95bc1885b9506dR19-R20

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+you may encounter one or more of the following issues:
+
+- Your application may not compile.
+- Your application may not be compatible with data that originated from an
```
This is scary.  What change is this referring to?

ccho-mongodb replied:
This is related to [JAVA-3854](https://jira.mongodb.org/browse/JAVA-3854). The issue report mentions incompatible serialization ids when attempting to use two different driver versions to serialize/deserialize an object that contains an ObjectId.
It's also what I was attempting to communicate (here)[https://github.com/mongodb/docs-java/pull/212#discussion_r824237100].

jyemin replied:
It just makes it sound a lot scarier than it actually is.  The serialization incompatibility in JAVA-3854 is not about serializing ObjectId instances to and from BSON, which would be absolutely horrible.  It's about serializing to and from Java's Object Serialization format, which is a much, much smaller group of people.  So I think including it here as a general statement will do more harm than good.

ccho-mongodb replied:
I know this has to do with Java's object serialization and agree it has to do with a smaller audience, hence my comment here: https://github.com/mongodb/docs-java/pull/212#discussion_r824845764

jyemin replied:
OK.  I still think it shouldn't be mentioned here, as it could easily be minconstrued.

ccho-mongodb replied:
I agree and removed it in this commit:
https://github.com/mongodb/docs-java/pull/212/commits/05fea111cd793956a1a46f862b9069aa8359dfde

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+
+.. _java-upgrade-from-legacy:
+
+Upgrade from the Legacy API
```
I don't think this belongs on an Upgrade Driver Versions page, as it's orthogonal to driver version: all driver versions support both legacy and new API.

Can we move to a new top-level page?

ccho-mongodb replied:
Sure, I can split it into its own page.

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+   * - `DBCollection.update() <{+api+}/apidocs/mongodb-driver-legacy/com/mongodb/DBCollection.html#update(com.mongodb.DBObject,com.mongodb.DBObject)>`__
+     - Use one of the following methods:
+
+       - `MongoCollection.updateOne() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/result/UpdateResult.html>`__
```
Also `replaceOne`

ccho-mongodb replied:
Thanks, will add that.

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+   * - Legacy
+     - Current
+
+   * - `MongoClientOptions <{+api+}/apidocs/mongodb-driver-legacy/com/mongodb/MongoClientOptions.html>`__
```
This could be blown out a lot more, as many of the names are different and MongoClientSettings is composed of other Settings classes (e.g. SocketSettings), whereas MongoClientOptions is flat.  Do you want to do that here or handle in a subsequent ticket?

ccho-mongodb replied:
If you can send me the info that you'd like to include, I could incorporate it in this ticket. Otherwise, if you need time to describe it fully, a new ticket would probably be best.

jyemin replied:
There are like 40 different settings.  I'm not actually sure how far we should go to enumerate them.  But fine to do it in a new ticket.

ccho-mongodb replied:
Created https://jira.mongodb.org/browse/DOCSP-21374

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+   * - `DBCollection <{+api+}/apidocs/mongodb-driver-legacy/com/mongodb/DBCollection.html>`__
+     - `MongoCollection<> <{+api+}/apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html>`__
+
+   * - `DBCursor <{+api+}/apidocs/mongodb-driver-legacy/com/mongodb/DBCursor.html>`__
```
This is incomplete.  In the legacy API, DBCursor takes on the role of both MongoCursor and FindIterable, whereas in the new API it's split. Perhaps we could give more detailed examples showing the difference.

ccho-mongodb replied:
Ok, I can add a note about that and ticket the work to create examples for each. Does that work?

ccho-mongodb replied:
Created: https://jira.mongodb.org/browse/DOCSP-21349

jyemin replied:
Sure

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,156 @@
+   :widths: 50 50
+   :class: compatibility-large
+
+   * - Legacy
```
The aggregate methods are quite different, though the names are the same.  Consider calling that out.

ccho-mongodb replied:
Ok, I'll need to look into that further. The only info I currently have is from your (forum post)[https://www.mongodb.com/community/forums/t/conversion-from-mongodb-driver-legacy-to-mongodb-driver-sync/125411] reported in the ticket.

ccho-mongodb commented:
Hi @jyemin , I've addressed your comments by splitting the page into one that addresses upgrading versions and another for migrating from the legacy API into two and creating the requested followup tickets. I may request another copy review from the docs team after I get your technical review approval.

jyemin commented on source/upgrade.txt:
```diff
@@ -0,0 +1,104 @@
+  to use a new format that minimizes serialization compatibility issues
+  across different versions of the driver.
+
+  Any objects that include an ``ObjectId`` field that were serialized with an
```
I'm still concerned about this. Many users will think we're talking about serializing into and out of MongoDB, not into and out of a Java ObjectInputStream/ObjectOutputStream.

Also, technically it's not the driver that actually throws an exception, it's the JDK

ccho-mongodb replied:
> I'm still concerned about this. Many users will think we're talking about serializing into and out of MongoDB, not into and out of a Java ObjectInputStream/ObjectOutputStream.

Ok, I'll remove it unless you have something specific you want to include.

ccho-mongodb commented on source/faq.txt:
```diff
@@ -99,29 +99,29 @@ What is a discriminator?
 
-Add the codec to the top your registry before the ``DefaultCodecRegister``
-and before the ``PojoCodecProvider``: 
+Add the codec to the top your registry before the default codec registry
```
I agree with the first change.
From the second part of the comment, I'm not sure whether you saw the sample code below that demonstrates it since it might be below the fold. I was originally just correcting an issue I noticed as I was going through some of the existing content. I spotted a couple other issues which I'll correct in the next commit, but just want to make sure I'm not missing anything from your comment.

ccho-mongodb commented:
@jyemin I've added the strict JSON format mode information, please take a look and review when you get the chance.

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -127,6 +132,17 @@ corresponds to the format of the example you want to see:
+
+         {
+           "_id:": { "$oid": "573a1391f29313caabcd9637" },
+           "createdAt": { "$date": "2020-09-30T18:22:51.648Z" },
```
This isn't what `STRICT` produces.  From my tests, this:

```
        System.out.println(doc.toJson(JsonWriterSettings.builder().outputMode(JsonMode.EXTENDED).build()));
        System.out.println(doc.toJson(JsonWriterSettings.builder().outputMode(JsonMode.RELAXED).build()));
        System.out.println(doc.toJson(JsonWriterSettings.builder().outputMode(JsonMode.STRICT).build()));
```

outputs

```
{"d": {"$date": {"$numberLong": "1647465431487"}}}
{"d": {"$date": "2022-03-16T21:17:11.487Z"}}
{"d": {"$date": 1647465431487}}
```

ccho-mongodb replied:
My test code was incorrect, will add the corrected version to the repo so future authors (hopefully) don't make the same mistake.

jyemin commented on source/upgrade.txt:
```diff
@@ -59,11 +59,6 @@ Version 4.2 Breaking Changes
   to use a new format that minimizes serialization compatibility issues
   across different versions of the driver.
 
-  Any objects that include an ``ObjectId`` field that were serialized with an
```
I think we should say something here, but just make it clear that it's Java Object Serialization that we are referring to and not BSON serialization.  Something like

Any Java object containing an ``Objectid` that have been serialized via Java Object Serialization[1] with an older version of the driver .... 

Note that this does _not_ effect ObjectId instances that have been serialized to BSON and stored in MongoDB.

[1] https://docs.oracle.com/javase/tutorial/jndi/objects/serial.html

jyemin reviewed (changes requested):
Looking good.  Just a few more small things.

jyemin reviewed (approved):
LGTM.  Thanks for working through all the details with me.
